### PR TITLE
Optimize transaction pool pruning

### PR DESCRIPTION
**Pull request by tomusdrw**
- Avoid calling into runtime when transaction is in the pool already
- Prevent revalidation of recently imported block extrinsics (they are becoming temporarily banned after import)

This is important for UTXO chains, cause after the transaction is imported we might have troubles figuring that it's no longer valid, since the inputs are already spent and pruned. Without this the transaction ends up in the `future` pool and is evicted only after some time.
CC @0x7CFE

### CODE REVIEW

File: core/transaction-pool/graph/src/pool.rs [hunk 1]
```diff
@@ -152,8 +152,51 @@ impl<B: ChainApi> Pool<B> {
...
+	pub fn prune(&self, at: &BlockId<B::Block>, parent: &BlockId<B::Block>, extrinsics: &[ExtrinsicFor<B>]) -> Result<(), B::Error> {
+		let mut tags = Vec::with_capacity(extrinsics.len());
+		// re use tags of known transactions to avoid calling into runtime
+		let ready = self.pool.read().by_hash(extrinsics.iter().map(|xt| self.api.hash(xt)));
+		let all = extrinsics.iter().zip(ready.clone());
+
+		for (xt, existing) in all {
+			match existing {
+				Some(tx) => {
+					tags.extend(tx.provides.iter().map(|x| x.clone()));
```
Comment (0x7CFE) [inline core/transaction-pool/graph/src/pool.rs]: Here `.map(|x| x.clone())` may be replaced with `.cloned()`

File: core/transaction-pool/graph/src/pool.rs [hunk 2]
```diff
@@ -152,8 +152,51 @@ impl<B: ChainApi> Pool<B> {
...
+	/// If it's known we take the provided tags we already have.
+	/// If it's not in the pool we query the runtime to get info about the tags.
+	/// Invalid extrinsics (most likely inherent) are silently ignored.
+	pub fn prune(&self, at: &BlockId<B::Block>, parent: &BlockId<B::Block>, extrinsics: &[ExtrinsicFor<B>]) -> Result<(), B::Error> {
+		let mut tags = Vec::with_capacity(extrinsics.len());
+		// re use tags of known transactions to avoid calling into runtime
+		let ready = self.pool.read().by_hash(extrinsics.iter().map(|xt| self.api.hash(xt)));
+		let all = extrinsics.iter().zip(ready.clone());
+
+		for (xt, existing) in all {
```
Comment (0x7CFE) [inline core/transaction-pool/graph/src/pool.rs]: I would be more than happy to see step-by-step comments describing the implementation. Unfortunately, it's really hard to follow the logic here. Also, I'd consider renaming `tx` and `xt` to something more verbose.

File: core/transaction-pool/graph/src/pool.rs [hunk 3]
```diff
@@ -152,8 +152,51 @@ impl<B: ChainApi> Pool<B> {
...
 	/// Prunes ready transactions that provide given list of tags.
-	pub fn prune_tags(&self, at: &BlockId<B::Block>, tags: impl IntoIterator<Item=Tag>) -> Result<(), B::Error> {
+	///
+	/// Optional list of included extrinsic hashes prevents them from being
+	/// validated or imported again.
+	pub fn prune_tags(
+		&self,
+		at: &BlockId<B::Block>,
+		tags: impl IntoIterator<Item=Tag>,
+		included: impl IntoIterator<Item=ExHash<B>> + Clone,
```
Comment (0x7CFE) [inline core/transaction-pool/graph/src/pool.rs]: Parameter name says nothing about it's purpose. Maybe consider renaming it?

File: core/transaction-pool/graph/src/ready.rs [hunk 4]
```diff
@@ -207,6 +207,14 @@ impl<Hash: hash::Hash + Member + Serialize, Ex> ReadyTransactions<Hash, Ex> {
 		self.ready.read().contains_key(hash)
 	}
 
+	/// Retrieve transaction by hash
+	pub fn by_hash(&self, hash: impl IntoIterator<Item = Hash>) -> Vec<Option<Arc<Transaction<Hash, Ex>>>> {
```
Comment (0x7CFE) [inline core/transaction-pool/graph/src/ready.rs]: Inconsistent formatting. Previously, associated type assignments were written without spaces.

File: core/transaction-pool/graph/src/rotator.rs [hunk 5]
```diff
@@ -58,11 +59,11 @@ impl<Hash: hash::Hash + Eq + Clone> PoolRotator<Hash> {
 	}
 
 	/// Bans given set of hashes.
-	pub fn ban(&self, now: &Instant, hashes: &[Hash]) {
+	pub fn ban(&self, now: &Instant, hashes: impl IntoIterator<Item = Hash>) {
```
Comment (0x7CFE) [inline core/transaction-pool/graph/src/rotator.rs]: same formatting issue as above

**Comment by 0x7CFE** [comment]
Sorry, hit the wrong button when submitting review. I didn't proposed any changes, just comments.

File: core/transaction-pool/graph/src/base_pool.rs [hunk 6]
```diff
@@ -226,6 +226,18 @@ impl<Hash: hash::Hash + Member + Serialize, Ex: ::std::fmt::Debug> BasePool<Hash
 		self.future.all()
 	}
 
+	/// Return ready transactions given list of hashes.
+	pub fn by_hash(&self, hashes: impl IntoIterator<Item=Hash> + Clone) -> Vec<Option<Arc<Transaction<Hash, Ex>>>> {
```
Comment (bkchr) [inline core/transaction-pool/graph/src/base_pool.rs]: Can we not use `hashes: Iter<Hash>`? Looking into the other functions, the hash is always used as reference. This could prevent a lot of extra allocations.

File: core/transaction-pool/graph/src/pool.rs [hunk 7]
```diff
@@ -152,9 +152,73 @@ impl<B: ChainApi> Pool<B> {
...
+	/// Used to clear the pool from transactions that were part of recently imported block.
+	/// To perform pruning we need the tags that each extrinsic provides and to avoid calling
+	/// into runtime too often we first lookup all extrinsics that are in the pool and get
+	/// their provided tags from there. Otherwise we query the runtime at the `parent` block.
+	pub fn prune(&self, at: &BlockId<B::Block>, parent: &BlockId<B::Block>, extrinsics: &[ExtrinsicFor<B>]) -> Result<(), B::Error> {
+		let mut tags = Vec::with_capacity(extrinsics.len());
+		// Get details of all extrinsics that are already in the pool
+		let in_pool = self.pool.read().by_hash(extrinsics.iter().map(|extrinsic| self.api.hash(extrinsic)));
+		// Zip the ones from the pool with the full list (we get pairs `(Extrinsic, Option<TransactionDetails>)`)
+		let all = extrinsics.iter().zip(in_pool.clone());
```
Comment (bkchr) [inline core/transaction-pool/graph/src/pool.rs]: Why do we need to `clone` here?

File: core/transaction-pool/graph/src/base_pool.rs (hunk 6 above)
Comment (tomusdrw) [reply core/transaction-pool/graph/src/base_pool.rs]: Do you mean `Iterator<Item = &'a Hash>`?
1. It adds quite some noise with lifetimes (and they are unforutnatelly required)
2. IT's not so easy to convert `Iterator<Item=Hash>` into `Iterator<Item=&Hash>` so we would need to accept `&[Hash]` as an input on the other function

File: core/transaction-pool/graph/src/pool.rs (hunk 7 above)
Comment (tomusdrw) [reply core/transaction-pool/graph/src/pool.rs]: Good catch, we can just use `iter()`

File: core/transaction-pool/graph/src/base_pool.rs (hunk 6 above)
Comment (tomusdrw) [reply core/transaction-pool/graph/src/base_pool.rs]: I've changed the logic though to just accept `&[Hash]` I think with current approach hashes were actually computed twice (if you look at `Clone` implementations of `Map` for instance it basically clones the input collection and applies the same Fn).
